Improve README.md for npm page and outsider (#346)
zestedesavoir/zmarkdown: packages/zmarkdown/README.md

# zmarkdown

  [![NPM Version][npm-image]][npm-url]
  [![NPM Downloads][downloads-image]][downloads-url]
  [![Linux Build][travis-image]][travis-url]
  [![Test Coverage][coveralls-image]][coveralls-url]

> This sub-repository is a **HTTP Server API** to provide fast and extensible **markdown parser**. It is the current Markdown engine powering [Zeste de Savoir][zds].

Reviewer: ```suggestion
This project is an **HTTP Server API** providing fast and extensible **markdown parser**. It is the Markdown engine powering [Zeste de Savoir][zds].
```

It is a collection of packages extending the [**remark**

> processor][processor] and its [**MDAST**][mdast] syntax tree, [**rehype**][rehype] (for HTML processing) and [**textr**][textr] (text transformation framework). It also provides [**MDAST**][mdast] to LaTeX compilation via [**rebber**][rebber] (and its [plugins][rebber-plugins]).

Reviewer: It is not a collection of packages, it's just a small express server leveraging remark, rehype, textr, and all of these: https://github.com/zestedesavoir/zmarkdown/tree/master/packages

#return: ["<p>Hello word</p>",{"disableToc":true,"languages":[],"depth":1},[]]
```

[npm-image]: https://img.shields.io/npm/v/zmarkdown.svg
[npm-url]: https://npmjs.org/package/zmarkdown
[downloads-image]: https://img.shields.io/npm/dm/zmarkdown.svg
[downloads-url]: https://npmjs.org/package/zmarkdown
[travis-image]: https://img.shields.io/travis/zestedesavoir/zmarkdown/master.svg?label=linux
[travis-url]: https://travis-ci.com/zestedesavoir/zmarkdown
[coveralls-image]: https://img.shields.io/coveralls/zestedesavoir/zmarkdown/master.svg
[coveralls-url]: https://coveralls.io/r/zestedesavoir/zmarkdown?branch=master

[zds]: https://zestedesavoir.com
[processor]: https://github.com/remarkjs/remark/blob/master/packages/remark
[mdast]: https://github.com/wooorm/mdast
[rehype]: https://github.com/rehypejs/rehype
[textr]: https://github.com/A/textr
[rebber]: https://github.com/zestedesavoir/zmarkdown/tree/master/packages/rebber#rebber--
[rebber-plugins]: https://github.com/zestedesavoir/zmarkdown/tree/master/packages/rebber-plugins#rebber-plugins--

## Features
 - Convert Markdown to HTML ;

>  - Convert Markdown to EPUB file ;

Reviewer: ```suggestion
 - Convert Markdown to EPUB compatible HTML ;
```

 - Convert Markdown to LaTeX ;
 - Convert Markdown to TEX file.

## Installation

This is a [Node.js](https://nodejs.org/en/) module available through the [npm registry][npm-url]. Install with npm:

```
npm install zmarkdown
```

> ## Getting start

Reviewer: ```suggestion
## Getting started
```

 1. Start your zmarkdown server  `npm run server`
 2. Send a `POST` request to `http://localhost:27272/{endpoint}`
 

> `pm2 monit`: provides a realtime dashboard that fits directly into your terminal, is a [simple way to monitor][pm2-monit] the ressource usage of you server.

Reviewer: ```suggestion
`pm2 monit`: provides a realtime dashboard that fits directly into your terminal, it is a [simple way to monitor][pm2-monit] the resource usage of you server.
```

[pm2-monit]:http://pm2.keymetrics.io/docs/usage/monitoring/

> ### Limit the ressource usage of your server

Reviewer: ```suggestion
### Limit the resource usage of your server
```

> You can change the number of thread (default: `3`) and the max-memory of each thread (default: `150M`) in you package.json to `scripts.server` :

Reviewer: ```suggestion
You can change the number of threads (default: `3`) and the max-memory of each thread (default: `150M`) in your package.json at `scripts.server` :
```

```json
    "server": "pm2 start -f server/index.js -i 3 --max-memory-restart 150M",
```

## Requests

> All endpoints respond to `HTTP POST`. The request body must be a JSON and `md` is required in every endpoints, `opts` object depends of endpoint, each endpoint have their own options.

Reviewer: ```suggestion
All endpoints respond to `HTTP POST`. The request body must be JSON with a required `md` key. An optional `opts` key can be provided, the value of which depends on the endpoint.
```

### URL

```
POST http://localhost:27272/{endpoint}
```

### Body

#### Required Body JSON Value

| Name | Type | Description |
| - | - | - |
| `md` | string | markdown source string. |

> #### Optionnal Body JSON Value

Reviewer: ```suggestion
#### Optional Body JSON Value
```

| Name | Type | Description |
| - | - | - |

> | `opts` | JSON | Options, specific option for each endpoints. This part will be supplemented in **Request** of the endpoints sections. |

Reviewer: ```suggestion
| `opts` | JSON | Options specific to this endpoint. This object is documented in the **Request** section of each endpoint. |
```

### Response

All endpoints return `[contents, metadata, messages]` as JSON.

| Name | Type | Description |
| - | - | - |
| `contents` | string | the rendered HTML or LaTeX. |

> | `metadata` | object | depends on request options. This part will be supplemented in **Response** of the endpoints sections. |

Reviewer: ```suggestion
| `metadata` | object | depends on request options. This object is documented in the **Response** section of each endpoint. |
```

| `messages` | string[] | info/debug/errors from parsers, plugins, compilers, etc. |

Only `metadata` is described in the **Response** sections below.

> ### Example

Reviewer: This section is a bit confusing, I suggest removing it. There already is an example with `curl` above.

Sending `POST` to `http://localhost:27272/html` a JSON:

```
{
  "md": "Hello word",
  "opts": {}
}
```

# Endpoints 

[ref-epub]: #epub
[ref-html]: #html
[ref-latex]: #latex
[ref-tex]: #tex

## epub

> Markdown to EPUB file

Reviewer: ```suggestion
Markdown to EPUB compatible HTML
```

| `opts.inline` | bool | default: `false`, Only parse inline Markdown elements (such as links and emphasis, unlike lists and fenced code blocks). |
| `opts.stats` | bool | default: `false`, Will compute and return statistics about markdown text. |

### Response `metadata` values

| Name | Type | Description |
| - | - | - |
| `metadata.disableToc` | bool | Whether or not the input Markdown did **not** contain headings (`#`, `##`, …). This property is named that way because we use it to disable Table of Contents generation when no headings were found.<br>`disableToc: true` means *no headings*<br>`disableToc: false` means at least one *heading* |
| `metadata.ping` | string[] | undefined if `opts.disable_ping: true`<br>The list of nicknames returned by `remark-ping`. Can be used to send "ping" notifications to the corresponding users.<br>Note: this is fully customizable, `remark-ping` can validate potential *ping*s by any means, including sending an HTTP request (we recommend `HEAD`) to a REST API to make sure this username actually exists. |
| `metadata.languages` | string[] | A list of unique languages used in GitHub Flavoured Markdown fences with a flag. |

> | `metadata.stats` | object | stats about the parsed text:<br>- `signs`: number of chars, spaces included.<br>- ` words`: number of words. |

Reviewer: A small but complete example (request/response) would be useful for this endpoint since it's the most useful one. Ideally something with >=2 fenced code blocks featuring different languages, >= 2 pings, etc.
